torch/jit/__init__.py

```diff
@@ -580,5 +570,77 @@ def _time(trace_name, name, time=True):
+    """
+    Verify that a JIT compiled model has the same behavior as its uncompiled
+    version along with its backwards pass.
+
+    This function has side-effects (e.g., it executes your model / saves and loads
+    parameters), so don't expect the model to come out exactly the same as what
+    you passed in.
+
+    Arguments:
+        model (compiled torch.nn.Module or function): the module/function to be
+            verified.  The module/function definition MUST have been decorated with
+            `@torch.jit.compile`.
+        args (tuple or Variable): the positional arguments to pass to the
+            compiled function/module to be verified.  A non-tuple is assumed to
+            be a single positional argument to be passed to the model.
+        loss_fn (function, optional): the loss function to be applied to
+            the output of the model, before backwards is invoked.  By default,
+            we assume that a model returns a single result, and we :func:`torch.sum`
+            before calling backwards; if this is inappropriate, you can pass your
+            own loss function.  Note that if a model returns a tuple of results,
+            these are passed as separate arguments to `loss_fn`.
+        devices (iterable of device IDs, optional): the GPU devices which the
+            compiled module will be run on.  This determines the RNG state we
+            must save when running both compiled and uncompiled versions of the model.
+    """
+    # TODO: In principle, we track device information in our trace, so it
+    # should be possible to check if our execution actually obeyed the 'devices'
+    # the user provided.
+
+    # TODO: Consider adding a utility function to torch.jit to test
+    # for this case
+    if not isinstance(model, _CompiledMixin):
+        raise TypeError("Cannot verify an uncompiled module.  Add @torch.jit.compile to compile it")
+
+    if not isinstance(args, tuple):
+        args = (args,)
+
+    saved_args = _clone_inputs(args)
+    saved_state = copy.deepcopy(model.state_dict())
+
+    def run_fwd_bwd(args, force_trace=False):
+        in_vars, _ = _flatten(args, model.state_dict(keep_vars=True).values())
+        # We use a special API to reset the trace and compile it from scratch.
+        out = model(*args, _force_trace=force_trace)
+        if not isinstance(out, tuple):
+            out = (out, )
+        out_vars, _ = _flatten(out)
+        saved_outs = [v.data.clone() for v in out_vars]
+        loss = loss_fn(*out)
+        grads = torch.autograd.grad([loss], in_vars)
+        # TODO: I'm not sure if the clone here is necessary but it is safer
+        saved_grads = [v.data.clone() for v in grads]
+        return (saved_outs, saved_grads)
+
+    with torch.random.fork_rng(devices, _caller="torch.jit.verify"):
+        uncompiled_outs, uncompiled_grads = run_fwd_bwd(args, force_trace=True)
+        assert model.has_trace_for(*args)
+
+    model.load_state_dict(saved_state)
```

Comment:
Uh this is annoying. After `torch.jit.verify` param data will point to a different thing. Why can't you just use `copy` to create a clone of the whole module tree, including Parameters and use this to compare against the original module

Reply:
I don't think this is true anymore. Since https://github.com/pytorch/pytorch/pull/451, `load_state_dict` [performs a `copy_` and not an assignment anymore](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L401), so I think this should be fine, right?

Reply:
Ah my bad. Thanks @fmassa

Reply:
To be honest, I wrote the code this way because I was not sure `copy` would always be supported on user-written Modules. If it could be broken in subtle ways, I would prefer not to use it, since it adds another way for things to fail (whereas `state_dict` and `load_state_dict` are blessed official APIs.) But if we do know deep copies always work, it is probably more robust to make a straight up copy at the beginning.